gossip/src/duplicate_shred_handler.rs

```diff
@@ -0,0 +1,451 @@
+use {
+    crate::{
+        cluster_info_duplicate_shred_listener::ClusterInfoDuplicateShredHandler,
+        duplicate_shred::{DuplicateShred, Error},
+    },
+    itertools::Itertools,
+    log::*,
+    solana_ledger::{
+        blockstore::Blockstore, blockstore_meta::DuplicateSlotProof,
+        leader_schedule_cache::LeaderScheduleCache, shred::Shred,
+    },
+    solana_sdk::{clock::Slot, pubkey::Pubkey},
+    std::{
+        collections::{HashMap, HashSet},
+        sync::Arc,
+    },
+};
+
+const CLEANUP_EVERY_N_LOOPS: usize = 10;
+// Normally num_chunks is 3, because there are two shreds (each is one packet)
+// and meta data. So we discard anything larger than 3 chunks.
+const MAX_NUM_CHUNKS: u8 = 3;
+// We only allow each pubkey to send proofs for 5 slots, because normally there
+// is only 1 person sending out duplicate proofs, 1 person is leader for 4 slots,
+// so we allow 5 here to limit the chunk map size.
+const ALLOWED_SLOTS_PER_PUBKEY: usize = 5;
+
+struct ProofChunkMap {
+    num_chunks: u8,
+    missing_chunks: u8,
```

Comment:
Given the `chunks` field, one of these 2 fields is redundant.

Reply:
I originally thought allowing num_chunks to be 1 or 2 just in case we change shred size to be larger (which I heard is somewhere in the distant future) or somehow we figured how to pass on duplicate proofs without actual shreds. Are you suggesting we stick to chunk of 3 here? Adding 1 byte shouldn't be that expensive.

Reply:
done